# EJERCICIO 1
Ejercicio 1 de evaluación de expresiones en Python
$\frac {\pi ^ 2}{6} = .\sum_{n=1}^{\infty} \frac{1}{n ^ 2}$

In [40]:
import math
res = math.pi ** 2 / 6
res2=0
for n in range(1,10):
    res2+=1 / (n**2)

print(f"valor teorico: {res}")
print(f"Aproximacion: {res2}")
error= abs(res-res2)
print(f"Error: {error}")


valor teorico: 1.6449340668482264
Aproximacion: 1.5397677311665408
Error: 0.10516633568168565


### Prioridad de los Operadores
1: Se eleva 2 al cuadrado
2: Se multiplica por 4
3: Se le suma 3
Total:19
En este caso, Python primero evalúa el operador de potencia ( **), luego la multiplicación, y finalmente la suma, siguiendo la jerarquía de operadores.

In [41]:
# Expresion
resultado = 3 + 4 * 2 ** 2
print(resultado) 
# ¿Cuál crees que es el resultado?

19


### EJERCICIO 1: EVALUACION DE EXPRESIONES DE PYTHON
Instrucciones: Analice las siguientes expresiones, anote las reducciones en un cuadernillo de Jupyter y compruebe el resultado en el lenguaje python.

##  1:expr1 = (3 + 2) * 4 
1-> Python evalua el operador () y todo lo que está dentro, por tanto: (3+2)=5
2-> Finalmente, siguiendo la jerarquía de operadores, el producto se multiplica por 4. Resultado: 5*4=20

## 2:expr2 = 2 * 3 ** 2 
1-> Siguiendo la jerarquía de operadores, se eleva al cuadrado el número 3. 
3 ** 2 = 9
2-> Posteriormente, se multiplica el resultado anterior por 2. Resultado: 9*2= 18

## 3:expr3 = 10 // 3 + 2
1-> Se realiza la división entera, ya que va primero en la jerarquía. Por tanto: 10 // 3 = 3. Ya que se redondea
2-> Finalmente, se realiza la suma. Resultado: 3+2=5

## 4:expr4 = 17 % 5 * 2
1-> Se realiza primero el operador %, ya que sin paréntesis está siguiendo la notación infija. El resultado es 2.
2-> Finalmente, el resultado se multiplica por dos. Resultado: 2*2=4

## 5:expr5 = -2 ** 2
1-> Dado que el signo negativo sólo afecta al primer número, simplemente se eleva al cuadrado. Resultado: -4

## 6:expr6 = 3 + 4 * 2 ** 2 // 3
1-> Se realiza la elevación a la potencia. 2 ** 2 = 4
Nuestra operación se simplifica a: 3+4*4//3
2-> Se realiza la multiplicación. 4*4=16
Nuestra operación se simplifica a: 3 + 16 // 3
3-> Se realiza la división entera de 16 // 3. Nuestra operación se simplifica a: 3+5
4-> Finalmente, se realiza la suma, que nos da de resultado 8

## 7:expr7 = 3 < 4 < 5
1-> Siguiendo la notación infija. El 3 < 4, por tanto es verdadero.
2-> 4 < 5, por tanto, es verdadero
Resultado final: La expresión es verdadera <true>


## 8:expr8 = (3 > 2) and (5 < 10) or False
1-> Si la afirmación 3>2 es verdadera, y también 5<10 es verdadera, nuestra expresión sería verdadera.
2-> En caso de no cumplirse la afirmación anterior, la afirmación puede ser falsa. En este caso, la afirmación 5>10 es falsa, por lo que nuestra expresión es falsa. Por lo que al entrar dentro del or <false>, nuestra afirmación es verdadera <true>

## 9: expr9 = 5 & 3 + 1
1-> Se suma 3+1. La expresión se ve de la siguiente manera: 5 & 4
2-> Dado que el operador & realiza un AND entre 5 y 4 en binario, realizaríamos el siguiente cálculo

AND
0 1 0 1  Cinco
0 1 0 0  Cuatro
-------
0 1 0 0   Resultado: 4


## 10:expr10 = (2 + 3) * 4 ** 2 // (5 % 3) and not False
1-> Se realiza primero la operación (2+3) y la operación (5%3). La expresión se ve así hasta este punto:
5 * 4 ** 2 // 2
2-> Se realiza la elevación al cuadrado. La expresión se ve así hasta este punto:
5 * 16 // 2
3-> Siguiendo la notación infija se multiplica 5*16 y finalmente se realiza la división entera entre 2. Resultado: 40
4-> Dado que pide que no sea falso, la afirmación es verdadera ya que no está realizando ninguna condición dentro de la expresión.

## 11: expr11 =  ((65 & 66) > 70 ) And not (65 ^ 66) == 0 
1-> Realizar la siguiente parte: ((65 & 66) > 70 )
Comenzamos realizando el AND
  AND
  0 1 1 1 1 1 0                        62
  1 0 0 0 1 0 0   
---------------------                  68
  1 0 0 0 1 0 0              Resultado:68
  El resultado debe ser mayor de 70. Dado que el resultado es 68, la expresión es falsa.
2-> Posteriormente, nos pide que también se cumpla la siguiente condición not (65 ^ 66) == 0, que nos dice que en un XOR entre 65 y 66 no puede dar 0:

XOR
  1 0 0 0 0 0 1                        65
  1 0 0 0 0 1 0                        66
---------------------                 
  0 0 0 0 0 1 1                   Resultado: 3
Dado que el resultado no es igual a 0, la afirmación es verdadera. Sin embargo, como no se cumplió la primera parte, nuestra expresión es falsa.

In [42]:
import math
expr1 = (3 + 2) * 4 
print(f"valor de la primera expresion: {expr1}")

expr2 = 2 * 3 ** 2 
print(f"valor de la segunda expresion: {expr2}")

expr3 = 10 // 3 + 2
print(f"valor de la tercera expresion: {expr3}")

expr4 = 17 % 5 * 2
print(f"valor de la cuarta expresion: {expr4}")

expr5 = -2 ** 2
print(f"valor de la quinta expresion: {expr5}")

expr6 = 3 + 4 * 2 ** 2 // 3
print(f"valor de la sexta expresion: {expr6}")

expr7 = 3 < 4 < 5
print(f"valor de la septima expresion: {expr7}")

expr8 = (3 > 2) and (5 < 10) or False
print(f"valor de la octava expresion: {expr8}")

expr9 = 5 & 3 + 1
print(f"valor de la novena expresion: {expr9}")

expr10 = (2 + 3) * 4 ** 2 // (5 % 3) and not False
print(f"valor de la decima expresion: {expr10}")

expr11 =  ((65 & 66) > 70 ) and not (65 ^ 66) == 0 
print(f"valor de la decima primer expresion: {expr11}")

valor de la primera expresion: 20
valor de la segunda expresion: 18
valor de la tercera expresion: 5
valor de la cuarta expresion: 4
valor de la quinta expresion: -4
valor de la sexta expresion: 8
valor de la septima expresion: True
valor de la octava expresion: True
valor de la novena expresion: 4
valor de la decima expresion: True
valor de la decima primer expresion: False


### EJERCICIO 2: EVALUACION DE EXPRESIONES DE JAVA
Instrucciones: Analice las siguientes expresiones, anote las reducciones y compruebe el resultado 
en el lenguaje. Si una expresión está mal redactada, favor de repararla. Se permite crear variables
extra. Puede utilizar esta herramienta Java Online Compiler.

<br> Nota: todas las expresiones comienzan con los siguientes valores:
<br>int x=10;
<br>int b=10;


## Expresión #1: System.out.println(x++);
La siguiente expresion imprime como resultado 10 debido al orden en el que se escribe la 
expresión. Primero imprime el valor de x (10) y posteriormente realiza el incremento.

## Expresión #2: System.out.println(++x);
En este caso sucede lo contrario a la expresión anterior, ya que primero incrementa x (ahora
vale 11) y posteriormente imprime el valor de x

## Expresión #3: System.out.println(b++ * -3 == -3 % x++);
La siguiente expresión es falsa dado que dice que si 10 multiplicado por -3 (-30) es igual
a -3 % 10 (que nos da un resultado de -3) es verdadero.
Nota: dado que el incremento está del lado derecho, se realiza después de la operación, por lo
que el incremento no afecta el cálculo.

## Expresión #4: 30 - 10 /= % ++b + 30 % 7 * 2
La expresión presenta errores de sintáxis, por lo que se tienen que separar ciertos datos en variables para recibir un resultado.
     <br>  double j= 30-10;
      <br> j/= ++b + (30 % 14);
       <br>System.out.println(j);
<br>Lo primero que se hizo fue realizar una variable adicional, ya que el operador /= requiere de algún valor para realizarlo. Posteriormente se eliminó el % ya que es inválida al no contar con algún valor a la izquierda para realizar el cálculo. Finalmente, se realizó la multiplicación de 7*2 para que se realizara el cálculo directamente con 30. 
<br>
<br>Explicación de j: Primero se realiza la operación 30 % 14. Resultado: 2
Posteriormente, a ++b (11) se le suma 2 (13)
<br>Finalmente, se divide j(20) entre el resultado (13)
<br>Resultado: 1.5384615384615385

## Expresión #5: x = -50 * x++ * 80 % b / 4 *= 8
        int x=10;
   <br>    int b=10;
   <br>     double j= -50 * x++ * 80 % b / 4;
   <br>     j *= 8;
   <br>     System.out.println(-50 * x++ * 80 % b / 4);
<br>
<br>
1> En la siguiente expresión  hay diversos errores, pero el primero que se identificó es que no tenía ninguna coherencia tener el x= ya que este ya almacena otro valor.
<br> -50 * x++ * 80 % b / 4 *= 8
<br> 2> Dado a que hay un *= y este requiere una variable para funcionar, se almacenaron los cálculos de la izquierda en una variable.
<br> double j= -50 * x++ * 80 % b / 4;
<br> En esta variable se realizó primero el módulo % (0). Al dar como resultado 0 y lo demás multiplicarse o dividirse por este valor, la ecuación es 0.
<br> 3> Finalmente, se realiza la parte faltante de la ecuación en j.
<br>         j *= 8;
  <br>       Resultado: 0


## Expresión #6:  System.out.println(b *= x++ - 8 + 3 - 5 / 4 % - 3 + 8);
<br>1> Se realiza primero la división de 5 entre 4.
<br>b *= x++ - 8 + 3 - 1 % - 3 + 8
<br>Se redondea a 1 porque no trae el punto decimal el 5 y el 4
<br>2> Se realiza el %
<br>b *= x++ - 8 + 3 - 1 + 8
<br>3>Se realizan las sumas y restas.
<br>b *= x++ + 2
<br>10 + 2= 12
<br>4> Finalmente, se realiza el *=
<br>10 *= 12 que es lo mismo que 10*12
<br>Resultado: 120

## Expresión #7: System.out.println(x = b * = 50 / 30 * ++x + 80);
Hay un error de sintáxis en el * =, por lo que se quitó el espacio para realizar el cálculo.
<br>x = b *= 50 / 30 * ++x + 80
<br>1> Se divide 50 entre 30.
<br>x = b *= 1 * ++x + 80
<br>El resultado es 1.6 pero se redondea a 1
<br>2> Se multiplica 1 por 11.
<br>x = b *= 11 + 80
<br>3> Se suma 11 + 80
<br>x= b *= 91
<br>4>Finalmente, se multiplica 10 por 91.
<br>Resultado: 910

## Expresión #8: ++x % 20.2 == -x /= b % = --x
Dado que la expresión busca comparar 2 expresiones, se generaron las variables j y k para realizar la comparación correctamente. 
<br>double j= ++x % 20.2;
<br>double k=-x;
<br>k/= b % --x;
<br> System.out.println(j == k);
<br>
<br>1> En la variable j, se realiza 11 % 20.2. Resultado: 11
<br>2> En la variable k se almacena un -10
<br>3> Se realiza b % --x y finalmente k se divide entre el resultado anterior. Resultado: Infinito. Esto se debe a que 10 % 10 es igual a 0 y el programa lo interpreta como infinito
<br>Dado a que j no vale lo mismo que k, la expresión es falsa.

## Expresión #9: 1 *= b++ * b-- / 3 % ++b *= 10 + 1;
Dado que la expresión contaba con un error al intentar ejecutar los operadores *= y %, se crearon las variables j y k para que la expresión se pudiera realizar.
<br>        int j=1;
<br>        j *= b++ * b-- / 3;
<br>        int k=++b;
<br>        k*=10+1;
<br>       System.out.println(j % k);

<br>1> Se iguala j igual a 1. Posteriormente, este valor se multiplica por 10, luego por 11 y finalmente se divide entre 3. Resultado: 36 ya que es una división entera.
<br>2>Se crea la variable k, que inicialmente va a valer 11 (b++), y lo multiplicaremos por 11. Resultado: 121
<br>3>Finalmente, realizamos j%k. Resultado: 36


## Expresión #10: -10 * -0 *= -10 *=x++ + x+=x*x ;
Este ejercicio presenta un error ya que no cuenta con una variable declarada para almacenar las multiplicaciones que desea realizar, por lo que se estructuró de la siguiente manera:
<br>int j= -10 * -0;
<br>        j*=-10;
<br>        j*=x++ + x;
<br>        j+= x*x;
<br>       System.out.println(j);
<br>1> Primero se multiplica -10*0. j equivale a 0
<br>2> Posteriormente,se multiplica por -10. j vale 0
<br>3> Ahora se multiplica j por la suma de 10 y 11. j vale 0
<br>4> A j(0) le sumamos la multiplicación de 11*11 (x)
<br>Resultado:121

## Expresión #11: a * b == b++ + b-- + --b + x++ + x--;
Dado que la variable a no existe, se eliminó junto con el operador de multiplicación.
       System.out.println( b == b++ + b-- + --b + x++ + x--);
<br>1>Se sustituyen los valores de la siguiente manera para evaluar si b es igual a la suma de los siguientes 
valores> b== 10 + 11(en la parte anterior se le incrementó 1) + 10 (En la anterior se le disminuyó 1) + 10 + 11 (en el anterior se le agregó 1). Resultado: 51
<br>2> Dado a que b no es igual a 51, la expresión es falsa.
